<a href="https://colab.research.google.com/github/y626010-prog/-Retail-Mall-/blob/main/Retail_Mall_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install faker

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 21.0 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
from faker import Faker
import random

In [ ]:
fake = Faker()

# Customers
customers = pd.DataFrame({
    "customer_id": range(1,101),
    "first_name": [fake.first_name() for _ in range(100)],
    "last_name": [fake.last_name() for _ in range(100)],
    "email": [fake.email() if random.random()>0.05 else None for _ in range(100)],
    "gender": [random.choice(["M","F"]) if random.random()>0.05 else None for _ in range(100)],
    "age": [random.randint(18,65) for _ in range(100)],
    "loyalty_member": [random.choice([True,False]) for _ in range(100)],
    "phone_number": [fake.phone_number() if random.random()>0.07 else None for _ in range(100)]
})
print(customers.info())
print("__________________")
customers.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   customer_id     100 non-null    int64 
 1   first_name      100 non-null    object
 2   last_name       100 non-null    object
 3   email           97 non-null     object
 4   gender          92 non-null     object
 5   age             100 non-null    int64 
 6   loyalty_member  100 non-null    bool  
 7   phone_number    92 non-null     object
dtypes: bool(1), int64(2), object(5)
memory usage: 5.7+ KB
None
__________________


,customer_id,first_name,last_name,email,gender,age,loyalty_member,phone_number
0,1,Shelly,Foster,bryantrichard@example.com,None,35,False,None
1,2,Matthew,Mcneil,jimmy20@example.com,F,30,True,+1-205-790-5011x7385
2,3,Craig,Wiggins,ashley40@example.net,F,46,True,(999)436-8753
3,4,Gregory,Ochoa,jacquelinenunez@example.com,M,52,True,4078844340
4,5,Tara,West,andersondanielle@example.org,M,32,False,(731)719-3020x60521


In [ ]:

# Branches
locations = ["Egypt", "UAE", "Saudi Arabia", "USA", "UK"]

branches = pd.DataFrame({
    "branch_id": range(1,6),
    "branch_name": [f"Mall_Branch_{i}" for i in range(1,6)],
    "location": [random.choice(locations) for _ in range(5)],
    "manager_name": [fake.name() if random.random()>0.05 else None for _ in range(5)],
    "open_since": [fake.date_between(start_date='-10y',end_date='-1y') for _ in range(5)]
})

print(branches.info())
print("__________________")
branches.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   branch_id     5 non-null      int64 
 1   branch_name   5 non-null      object
 2   location      5 non-null      object
 3   manager_name  5 non-null      object
 4   open_since    5 non-null      object
dtypes: int64(1), object(4)
memory usage: 332.0+ bytes
None
__________________


,branch_id,branch_name,location,manager_name,open_since
0,1,Mall_Branch_1,Egypt,John Roberts,2023-01-01
1,2,Mall_Branch_2,USA,Kyle Sullivan,2023-10-18
2,3,Mall_Branch_3,UAE,Tony Mccarthy,2017-08-20
3,4,Mall_Branch_4,Saudi Arabia,Mr. Francisco Oliver,2016-05-11
4,5,Mall_Branch_5,USA,Sheryl Moore,2023-04-19


In [ ]:
fake = Faker()

brand_category = {
    "Electronics": ["Samsung","Apple","Sony","LG","HP","Lenovo"],
    "Clothing": ["Zara","H&M"],
    "Shoes": ["Nike","Adidas","Puma"],
    "Accessories": ["Rolex"],
    "Home": ["LG","Samsung"]
}

product_names = {
    "Electronics": ["Smart TV","Laptop","Smartphone","Tablet","Headphones","Bluetooth Speaker"],
    "Clothing": ["T-Shirt","Jeans","Jacket","Dress","Hoodie","Sweater"],
    "Shoes": ["Running Shoes","Sneakers","Boots","Sandals"],
    "Accessories": ["Watch","Sunglasses","Wallet","Backpack"],
    "Home": ["Microwave","Vacuum Cleaner","Coffee Maker","Blender"]
}

categories = list(brand_category.keys())

product_data = []

for i in range(1001,1101):

    category = random.choice(categories)
    brand = random.choice(brand_category[category])
    name = random.choice(product_names[category])

    product_name = f"{brand} {name}"

    product_data.append({
        "product_id": i,
        "product_name": product_name,
        "category": category,
        "brand": brand,
        "cost_price": round(random.uniform(50,500),2),
        "stock_quantity": random.randint(0,500)
    })

products = pd.DataFrame(product_data)
products.info()
products.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   product_id      100 non-null    int64  
 1   product_name    100 non-null    object 
 2   category        100 non-null    object 
 3   brand           100 non-null    object 
 4   cost_price      100 non-null    float64
 5   stock_quantity  100 non-null    int64  
dtypes: float64(1), int64(2), object(3)
memory usage: 4.8+ KB


,product_id,product_name,category,brand,cost_price,stock_quantity
0,1001,Adidas Boots,Shoes,Adidas,186.18,242
1,1002,Rolex Watch,Accessories,Rolex,361.99,315
2,1003,Rolex Wallet,Accessories,Rolex,153.53,239
3,1004,Rolex Backpack,Accessories,Rolex,238.11,133
4,1005,LG Vacuum Cleaner,Home,LG,366.22,318


In [ ]:
# Sales
sales = pd.DataFrame({
    "sale_id": range(1,201),
    "customer_id": np.random.choice(customers["customer_id"],200),
    "branch_id": np.random.choice(branches["branch_id"],200),
    "sale_date": [fake.date_between(start_date='-1y',end_date='today') for _ in range(200)]
})
sales.info()
sales.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sale_id      200 non-null    int64 
 1   customer_id  200 non-null    int64 
 2   branch_id    200 non-null    int64 
 3   sale_date    200 non-null    object
dtypes: int64(3), object(1)
memory usage: 6.4+ KB


,sale_id,customer_id,branch_id,sale_date
0,1,75,5,2025-05-09
1,2,13,4,2026-01-19
2,3,66,2,2025-09-13
3,4,78,5,2025-12-20
4,5,71,4,2025-06-23


In [ ]:
# Sales Items
sales_items_list = []
item_id = 1
for sale in sales["sale_id"]:
    items_count = random.randint(1,4)
    for _ in range(items_count):
        product_id = random.choice(products["product_id"].tolist())
        cost = products.loc[products["product_id"]==product_id,"cost_price"].values[0]
        # unit_price منطقي أكبر من التكلفة
        unit_price = round(cost * random.uniform(1.2,1.8),2) if random.random()>0.02 else None
        quantity = random.randint(1,5) if random.random()>0.02 else None
        sales_items_list.append({
            "item_id": item_id,
            "sale_id": sale,
            "product_id": product_id,
            "quantity": quantity,
            "unit_price": unit_price
        })
        item_id += 1
sales_items = pd.DataFrame(sales_items_list)


sales_items.info()
sales_items.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511 entries, 0 to 510
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   item_id     511 non-null    int64  
 1   sale_id     511 non-null    int64  
 2   product_id  511 non-null    int64  
 3   quantity    499 non-null    float64
 4   unit_price  501 non-null    float64
dtypes: float64(2), int64(3)
memory usage: 20.1 KB


,item_id,sale_id,product_id,quantity,unit_price
0,1,1,1093,2.0,275.13
1,2,1,1086,4.0,568.38
2,3,1,1011,5.0,235.28
3,4,2,1071,5.0,319.04
4,5,2,1011,5.0,206.45


In [ ]:
sales_items["total_amount"] = sales_items["unit_price"] *sales_items["quantity"]

In [ ]:
sale_totals = sales_items.groupby("sale_id")["total_amount"].sum().reset_index()

In [ ]:
sales_with_total = sales.merge(sale_totals, on="sale_id", how="left")


In [ ]:
sales_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511 entries, 0 to 510
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   item_id       511 non-null    int64  
 1   sale_id       511 non-null    int64  
 2   product_id    511 non-null    int64  
 3   quantity      499 non-null    float64
 4   unit_price    501 non-null    float64
 5   total_amount  489 non-null    float64
dtypes: float64(3), int64(3)
memory usage: 24.1 KB


In [ ]:
#payments
payments = pd.DataFrame({
    "payment_id": range(1, len(sales_with_total)+1),
    "sale_id": sales_with_total["sale_id"],
    "payment_status": [random.choice(["Completed","Pending","Failed"]) if random.random()>0.08 else None for _ in range(len(sales_with_total))],
    "payment_method": [random.choice(["Credit Card","Cash","Digital Wallet"]) for _ in range(len(sales_with_total))],
    "amount_paid": sales_with_total["total_amount"],
    "payment_date": sales_with_total["sale_date"]
})
payments.info()
payments.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   payment_id      200 non-null    int64  
 1   sale_id         200 non-null    int64  
 2   payment_status  188 non-null    object 
 3   payment_method  200 non-null    object 
 4   amount_paid     200 non-null    float64
 5   payment_date    200 non-null    object 
dtypes: float64(1), int64(2), object(3)
memory usage: 9.5+ KB


,payment_id,sale_id,payment_status,payment_method,amount_paid,payment_date
0,1,1,Failed,Digital Wallet,320.58,2025-04-27
1,2,2,Pending,Cash,937.10,2025-07-23
2,3,3,Failed,Digital Wallet,2086.81,2026-02-21
3,4,4,None,Digital Wallet,2907.37,2025-11-21
4,5,5,Pending,Digital Wallet,2701.33,2026-03-08


In [ ]:

# Promotions
promotions = pd.DataFrame({
    "promo_id": range(1,11),
    "promo_name": [f"Season_Promo_{i}" for i in range(1,11)],
    "discount_percentage": [round(random.uniform(5,30),2) for _ in range(10)]
})

promotions.info()
promotions.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   promo_id             10 non-null     int64  
 1   promo_name           10 non-null     object 
 2   discount_percentage  10 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 372.0+ bytes


,promo_id,promo_name,discount_percentage
0,1,Season_Promo_1,22.10
1,2,Season_Promo_2,23.48
2,3,Season_Promo_3,9.67
3,4,Season_Promo_4,26.30
4,5,Season_Promo_5,12.07


In [ ]:
import pandas as pd
import numpy as np

# مثال على توليد بيانات بدون تكرار للـcombo
# نفترض إن promotions و products موجودين عندك

# توليد كل التركيبات الممكنة بين promo_id و product_id
all_combinations = pd.MultiIndex.from_product(
    [promotions["promo_id"], products["product_id"]], names=["promo_id", "product_id"]
).to_frame(index=False)

# نختار عشوائي 50 صف فريد
promotion_product = all_combinations.sample(50, random_state=42).reset_index(drop=True)

promotion_product.info()
promotion_product.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   promo_id    50 non-null     int64
 1   product_id  50 non-null     int64
dtypes: int64(2)
memory usage: 932.0 bytes


,promo_id,product_id
0,6,1022
1,8,1038
2,8,1041
3,7,1061
4,5,1012


CLEANING OF DATA

In [ ]:
#customer
customers.duplicated().sum()

In [ ]:
customers.isnull().sum()

In [ ]:
customers[customers["gender"].isnull()]

In [ ]:
customers.loc[(71),"gender"]="M"
customers['gender'].fillna(customers['gender'].mode()[0], inplace=True)

In [ ]:
customers.fillna({"email": "unknown"}, inplace=True)
customers.fillna({"phone_number": "unknown"}, inplace=True)
customers.isnull().sum()

,0
customer_id,0
first_name,0
last_name,0
email,0
gender,0
age,0
loyalty_member,0
phone_number,0


In [ ]:
# Branches
branches.isnull().sum()

In [ ]:
branches.info()

In [ ]:
branches.duplicated().sum()

In [ ]:
branches['open_since'] = pd.to_datetime(branches['open_since'])

In [ ]:
#products
products.info()

In [ ]:
products.duplicated().sum()

NameError: name 'products' is not defined

In [ ]:
#sales
sales.info()

In [ ]:
sales.duplicated().sum()

np.int64(0)

In [ ]:
sales['sale_date'] = pd.to_datetime(sales['sale_date'])

In [ ]:
sales.info()

In [ ]:
sales_items.duplicated().sum()

np.int64(0)

In [ ]:
sales_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 497 entries, 0 to 496
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   item_id       497 non-null    int64  
 1   sale_id       497 non-null    int64  
 2   product_id    497 non-null    int64  
 3   quantity      497 non-null    int64  
 4   unit_price    497 non-null    float64
 5   total_amount  497 non-null    float64
dtypes: float64(2), int64(4)
memory usage: 23.4 KB


In [ ]:
# حساب الربعيات
Q1 = sales_items['unit_price'].quantile(0.25)
Q3 = sales_items['unit_price'].quantile(0.75)

# حساب IQR
IQR = Q3 - Q1

# تحديد الحدود
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR


outliers = sales_items[
    (sales_items['unit_price'] < lower_bound) |
    (sales_items['unit_price'] > upper_bound)
]

print("Number of Outliers:", len(outliers))


Number of Outliers: 0


In [ ]:
sales_items['quantity'].fillna(sales_items['quantity'].mean(), inplace=True)
sales_items['unit_price'].fillna(sales_items['unit_price'].mean(), inplace=True)
sales_items.isnull().sum()

/tmp/ipykernel_3107/4138762360.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  sales_items['quantity'].fillna(sales_items['quantity'].mean(), inplace=True)
/tmp/ipykernel_3107/4138762360.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method

,0
item_id,0
sale_id,0
product_id,0
quantity,0
unit_price,0
total_amount,22


In [ ]:
sales_items['quantity'] = sales_items['quantity'].astype(int)

sales_items['total_amount'] = sales_items['quantity'] * sales_items['unit_price']
sales_items.head()


,item_id,sale_id,product_id,quantity,unit_price,total_amount
0,1,1,1093,2,275.13,550.26
1,2,1,1086,4,568.38,2273.52
2,3,1,1011,5,235.28,1176.40
3,4,2,1071,5,319.04,1595.20
4,5,2,1011,5,206.45,1032.25


In [ ]:
sales_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511 entries, 0 to 510
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   item_id       511 non-null    int64  
 1   sale_id       511 non-null    int64  
 2   product_id    511 non-null    int64  
 3   quantity      511 non-null    int64  
 4   unit_price    511 non-null    float64
 5   total_amount  511 non-null    float64
dtypes: float64(2), int64(4)
memory usage: 24.1 KB


In [ ]:
payments.info()

In [ ]:
payments['payment_date'] = pd.to_datetime(payments['payment_date'])

In [ ]:
payments['payment_status'].fillna(payments['payment_status'].mode()[0], inplace=True)
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   payment_id      200 non-null    int64  
 1   sale_id         200 non-null    int64  
 2   payment_status  200 non-null    object 
 3   payment_method  200 non-null    object 
 4   amount_paid     200 non-null    float64
 5   payment_date    200 non-null    object 
dtypes: float64(1), int64(2), object(3)
memory usage: 9.5+ KB


/tmp/ipykernel_8535/4159860220.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  payments['payment_status'].fillna(payments['payment_status'].mode()[0], inplace=True)


In [ ]:

promotion_product.duplicated().sum()

In [ ]:
promotion_product = promotion_product.drop_duplicates()

promotion_product.duplicated().sum()

In [ ]:

tables = {
    "products": products,
    "customers": customers,
    "sales": sales,
    "sales_items": sales_items,
    "promotions": promotions,
    "payments": payments,
    "branches": branches,
    "promotion_product": promotion_product
}


for name, df in tables.items():
    print(f"{'='*30}")
    print(f"Table Name: {name.upper()}")
    print(f"{'='*30}")

    print("\n--- INFO ---")
    print(df.info())

    print("\n--- FIRST 5 ROWS ---")
    display(df.head())
    print("\n")

Table Name: PRODUCTS

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   product_id      100 non-null    int64  
 1   product_name    100 non-null    object 
 2   category        100 non-null    object 
 3   brand           100 non-null    object 
 4   cost_price      100 non-null    float64
 5   stock_quantity  100 non-null    int64  
dtypes: float64(1), int64(2), object(3)
memory usage: 4.8+ KB
None

--- FIRST 5 ROWS ---


,product_id,product_name,category,brand,cost_price,stock_quantity
0,1001,H&M Dress,Clothing,H&M,139.46,98
1,1002,Rolex Backpack,Accessories,Rolex,231.12,132
2,1003,Samsung Coffee Maker,Home,Samsung,183.86,267
3,1004,Apple Headphones,Electronics,Apple,137.53,170
4,1005,Zara T-Shirt,Clothing,Zara,337.59,154




Table Name: CUSTOMERS

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   customer_id     100 non-null    int64 
 1   first_name      100 non-null    object
 2   last_name       100 non-null    object
 3   email           100 non-null    object
 4   gender          100 non-null    object
 5   age             100 non-null    int64 
 6   loyalty_member  100 non-null    bool  
 7   phone_number    100 non-null    object
dtypes: bool(1), int64(2), object(5)
memory usage: 5.7+ KB
None

--- FIRST 5 ROWS ---


,customer_id,first_name,last_name,email,gender,age,loyalty_member,phone_number
0,1,Jason,Weeks,gwilliams@example.com,F,57,True,2697806056
1,2,Laura,Crawford,brian52@example.org,F,59,True,(482)939-0140
2,3,Melinda,Mcdonald,ggarcia@example.net,M,38,False,626-255-3564x91175
3,4,Judy,Todd,robertgriffin@example.net,F,18,True,(294)773-5032x5868
4,5,Hannah,Williams,cdoyle@example.net,F,45,False,890.802.5470x8033




Table Name: SALES

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   sale_id      200 non-null    int64         
 1   customer_id  200 non-null    int64         
 2   branch_id    200 non-null    int64         
 3   sale_date    200 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(3)
memory usage: 6.4 KB
None

--- FIRST 5 ROWS ---


,sale_id,customer_id,branch_id,sale_date
0,1,16,2,2025-04-27
1,2,49,3,2025-07-23
2,3,43,4,2026-02-21
3,4,4,3,2025-11-21
4,5,53,2,2026-03-08




Table Name: SALES_ITEMS

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 509 entries, 0 to 508
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   item_id       509 non-null    int64  
 1   sale_id       509 non-null    int64  
 2   product_id    509 non-null    int64  
 3   quantity      509 non-null    int64  
 4   unit_price    509 non-null    float64
 5   total_amount  509 non-null    float64
dtypes: float64(2), int64(4)
memory usage: 24.0 KB
None

--- FIRST 5 ROWS ---


,item_id,sale_id,product_id,quantity,unit_price,total_amount
0,1,1,1028,1,320.580000,320.58000
1,2,2,1052,3,405.113293,1215.33988
2,3,2,1075,5,187.420000,937.10000
3,4,3,1030,1,722.040000,722.04000
4,5,3,1091,2,612.620000,1225.24000




Table Name: PROMOTIONS

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   promo_id             10 non-null     int64  
 1   promo_name           10 non-null     object 
 2   discount_percentage  10 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 372.0+ bytes
None

--- FIRST 5 ROWS ---


,promo_id,promo_name,discount_percentage
0,1,Season_Promo_1,20.23
1,2,Season_Promo_2,8.96
2,3,Season_Promo_3,13.96
3,4,Season_Promo_4,28.83
4,5,Season_Promo_5,11.41




Table Name: PAYMENTS

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   payment_id      200 non-null    int64         
 1   sale_id         200 non-null    int64         
 2   payment_status  200 non-null    object        
 3   payment_method  200 non-null    object        
 4   amount_paid     200 non-null    float64       
 5   payment_date    200 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 9.5+ KB
None

--- FIRST 5 ROWS ---


,payment_id,sale_id,payment_status,payment_method,amount_paid,payment_date
0,1,1,Pending,Cash,320.58,2025-04-27
1,2,2,Failed,Digital Wallet,937.10,2025-07-23
2,3,3,Failed,Credit Card,2086.81,2026-02-21
3,4,4,Completed,Cash,2907.37,2025-11-21
4,5,5,Completed,Cash,2701.33,2026-03-08




Table Name: BRANCHES

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   branch_id     5 non-null      int64         
 1   branch_name   5 non-null      object        
 2   location      5 non-null      object        
 3   manager_name  5 non-null      object        
 4   open_since    5 non-null      datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 332.0+ bytes
None

--- FIRST 5 ROWS ---


,branch_id,branch_name,location,manager_name,open_since
0,1,Mall_Branch_1,Egypt,Kevin Johnson,2021-04-04
1,2,Mall_Branch_2,Saudi Arabia,Donald Davis,2019-06-12
2,3,Mall_Branch_3,UK,Austin Calhoun,2016-06-02
3,4,Mall_Branch_4,USA,Felicia Rogers,2019-05-01
4,5,Mall_Branch_5,UAE,Kristopher Hawkins,2019-06-24




Table Name: PROMOTION_PRODUCT

--- INFO ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   promo_id    50 non-null     int64
 1   product_id  50 non-null     int64
dtypes: int64(2)
memory usage: 932.0 bytes
None

--- FIRST 5 ROWS ---


,promo_id,product_id
0,6,1022
1,8,1038
2,8,1041
3,7,1061
4,5,1012


CONVERTING DATA TO CSV FILE


In [ ]:
tables = {
    "products": products,
    "customers": customers,
    "sales": sales,
    "sales_items": sales_items,
    "promotions": promotions,
    "payments":payments,
    "branches":branches,
    "promotion_product":promotion_product
}

for name, df in tables.items():
    df.to_csv(f"{name}.csv", index=False)

In [ ]:
import os

folder = "data"
os.makedirs(folder, exist_ok=True)

for name, df in tables.items():
    df.to_csv(f"{folder}/{name}.csv", index=False)

In [ ]:
!pip install pyodbc
!pip install sqlalchemy

In [ ]:
!pip install xlsxwriter

In [ ]:
customers.info()

In [ ]:
promotion_product.duplicated().sum()

python analysis

In [ ]:

from google.colab import files
uploaded = files.upload() # سيطلب منك اختيار الملف من جهازك

In [ ]:
import os
print(os.listdir())

In [ ]:
import pandas as pd

import pandas as pd

# تأكد من كتابة الاسم بدقة كما هو (بما في ذلك كل النقط)
file_path = 'mull.csv.xls.csv.xlsx'

try:
    df = pd.read_excel(file_path)
    print("تم تحميل الملف بنجاح!")
    print(df.head())
except FileNotFoundError:
    print("الملف غير موجود بهذا الاسم، تأكد من رفعه (Upload) إلى البيئة التي تعمل عليها.")
except Exception as e:
    print(f"حدث خطأ آخر: {e}")


pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

df.head()

الملف غير موجود بهذا الاسم، تأكد من رفعه (Upload) إلى البيئة التي تعمل عليها.


,promo_id,product_id
0,6,1022
1,8,1038
2,8,1041
3,7,1061
4,5,1012


In [ ]:
df.isnull().sum()

In [ ]:
df['promo_name'] = df['promo_name'].fillna('No Promotion')
df['discount_percentage'] = df['discount_percentage'].fillna(0)

print(df.isnull().sum())

In [ ]:
df.info()

In [ ]:
df.duplicated().sum()
df[df.duplicated()]

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px


# Gender distribution
print("=== Gender ===")
print(df['gender'].value_counts())


In [ ]:
fig = px.bar(df['gender'].value_counts().reset_index(),
             x='gender', y='count',
             title='Gender Distribution',
             color='gender',
             color_discrete_map={'F': 'pink', 'M': 'steelblue'})
fig.show()

In [ ]:
# Age analysis
print("\n=== Age ===")
print(f"Mean: {df['age'].mean():.1f}")
print(f"Min: {df['age'].min()}")
print(f"Max: {df['age'].max()}")

In [ ]:
fig4 = px.histogram(df, x='age',
                    title='Age Distribution',
                    nbins=20,
                    color_discrete_sequence=['steelblue'])
fig4.show()

In [ ]:
# 1. Most purchasing age group
age_bins = [18, 25, 35, 45, 55, 65]
age_labels = ['18-25', '26-35', '36-45', '46-55', '56-65']
df['age_group'] = pd.cut(df['age'], bins=age_bins, labels=age_labels)

fig5 = px.bar(df['age_group'].value_counts().reset_index(),
              x='age_group', y='count',
              title='Most Purchasing Age Group',
              color='age_group')
fig5.show()

In [ ]:
# Payment Method
fig2 = px.bar(df['payment_method'].value_counts().reset_index(),
              x='payment_method', y='count',
              title='Payment Method Distribution',
              color='payment_method')
fig2.show()

In [ ]:
# 2. Payment method by gender
payment_gender = df.groupby(['gender', 'payment_method']).size().reset_index(name='count')

fig6 = px.bar(payment_gender,
              x='payment_method', y='count',
              color='gender',
              barmode='group',
              title='Payment Method by Gender',
              color_discrete_map={'F': 'pink', 'M': 'steelblue'})
fig6.show()

In [ ]:
# Location distribution
print("\n=== Location ===")
print(df['location'].value_counts())

In [ ]:
# Location Distribution
fig3 = px.bar(df['location'].value_counts().reset_index(),
              x='location', y='count',
              title='Location Distribution',
              color='location')
fig3.show()

In [ ]:
# Average spending + purchase count per customer
customer_analysis = df.groupby('customer_full_name').agg(
    avg_spending=('amount_paid', 'mean'),
    purchase_count=('sale_id', 'nunique')
).reset_index()

customer_analysis.columns = ['customer', 'avg_spending', 'purchase_count']
customer_analysis = customer_analysis.sort_values('avg_spending', ascending=False).head(10)

fig7 = px.bar(customer_analysis,
              x='customer', y='avg_spending',
              title='Top 10 Customers by Average Spending',
              color='avg_spending',
              color_continuous_scale='Blues',
              hover_data={'purchase_count': True})
fig7.show()

In [ ]:
customer_analysis = df.groupby('customer_full_name').agg(
    total_spending=('amount_paid', 'sum'),
    purchase_count=('sale_id', 'nunique')
).reset_index()

customer_analysis.columns = ['customer', 'total_spending', 'purchase_count']
customer_analysis = customer_analysis.sort_values('total_spending', ascending=False).head(10)

fig7 = px.bar(customer_analysis,
              x='customer', y='total_spending',
              title='Top 10 Customers by Total Spending',
              color='total_spending',
              color_continuous_scale='Blues',
              hover_data={'purchase_count': True})
fig7.show()

In [ ]:
# Customer Segmentation by Purchase Frequency
customer_freq = df.groupby('customer_full_name')['sale_id'].nunique().reset_index()
customer_freq.columns = ['customer', 'purchase_count']

# Segmentation
def segment(count):
    if count >= 4:
        return 'Loyal'
    elif count >= 2:
        return 'Regular'
    else:
        return 'One-Time'

customer_freq['segment'] = customer_freq['purchase_count'].apply(segment)

fig8 = px.pie(customer_freq['segment'].value_counts().reset_index(),
              names='segment', values='count',
              title='Customer Segmentation by Purchase Frequency',
              color_discrete_map={'Loyal': 'darkblue',
                                  'Regular': 'steelblue',
                                  'One-Time': 'lightblue'})
fig8.show()

In [ ]:
# Merge segment with gender
customer_gender = df[['customer_full_name', 'gender']].drop_duplicates()
customer_seg = customer_freq.merge(customer_gender,
                                    left_on='customer',
                                    right_on='customer_full_name')

fig9 = px.histogram(customer_seg, x='segment', color='gender',
                    barmode='group',
                    title='Customer Segmentation by Gender',
                    color_discrete_map={'F': 'pink', 'M': 'steelblue'})
fig9.show()

In [ ]:
# 1. Top selling categories
fig10 = px.bar(df['category'].value_counts().reset_index(),
               x='category', y='count',
               title='Sales by Category',
               color='category')
fig10.show()

In [ ]:
# 3. Revenue by category
revenue_category = df.groupby('category')['total_amount'].sum().reset_index()

fig12 = px.pie(revenue_category,
               names='category', values='total_amount',
               title='Revenue by Category')
fig12.show()

In [ ]:
# 2. Top 10 best selling products
fig11 = px.bar(df['product_name'].value_counts().reset_index().head(10),
               x='product_name', y='count',
               title='Top 10 Best Selling Products',
               color='count',
               color_continuous_scale='Blues')
fig11.show()

In [ ]:
# Top Brands by Sales Count
fig13 = px.bar(df['brand'].value_counts().reset_index().head(10),
               x='brand', y='count',
               title='Top 10 Brands by Sales Count',
               color='count',
               color_continuous_scale='Blues')
fig13.show()

In [ ]:
# Brand popularity by Category
brand_category = df.groupby(['brand', 'category']).size().reset_index(name='count')
brand_category = brand_category[brand_category['brand'].isin(df['brand'].value_counts().head(5).index)]

fig14 = px.bar(brand_category,
               x='brand', y='count',
               color='category',
               title='Top 5 Brands by Category',
               barmode='group')
fig14.show()

In [ ]:
# 1. Promotions usage vs No Promotion
promo_usage = df['promo_name'].apply(lambda x: 'With Promotion' if x != 'No Promotion' else 'No Promotion').value_counts().reset_index()

fig15 = px.pie(promo_usage,
               names='promo_name', values='count',
               title='Promotion Usage',
               color_discrete_map={'With Promotion': 'steelblue', 'No Promotion': 'lightblue'})
fig15.show()

In [ ]:
# 2. Most used promotions
promo_count = df[df['promo_name'] != 'No Promotion']['promo_name'].value_counts().reset_index().head(10)

fig16 = px.bar(promo_count,
               x='promo_name', y='count',
               title='Most Used Promotions',
               color='count',
               color_continuous_scale='Blues')
fig16.show()

In [ ]:
# 3. Age group response to promotions
promo_age = df[df['promo_name'] != 'No Promotion'].groupby('age_group').size().reset_index(name='count')

fig17 = px.bar(promo_age,
               x='age_group', y='count',
               title='Promotion Response by Age Group',
               color='age_group')
fig17.show()

In [ ]:
# Promotion Response by Gender - With vs Without
promo_gender = df.groupby(['gender', 'promo_name']).size().reset_index(name='count')
promo_gender['promo_status'] = promo_gender['promo_name'].apply(
    lambda x: 'With Promotion' if x != 'No Promotion' else 'No Promotion')

promo_gender_grouped = promo_gender.groupby(['gender', 'promo_status'])['count'].sum().reset_index()

fig17 = px.bar(promo_gender_grouped,
               x='gender', y='count',
               color='promo_status',
               barmode='group',
               title='Promotion Usage by Gender',
               color_discrete_map={'With Promotion': 'steelblue', 'No Promotion': 'lightblue'})
fig17.show()

In [ ]:
# Promotion by Category
promo_category = df[df['promo_name'] != 'No Promotion'].groupby('category').size().reset_index(name='count')

fig18 = px.bar(promo_category,
               x='category', y='count',
               title='Promotion Usage by Category',
               color='count',
               color_continuous_scale='Blues')
fig18.show()

In [ ]:
# Average discount by category
avg_discount = df[df['discount_percentage'] != 0].groupby('category')['discount_percentage'].mean().reset_index()

fig19 = px.bar(avg_discount,
               x='category', y='discount_percentage',
               title='Average Discount % by Category',
               color='discount_percentage',
               color_continuous_scale='Blues')
fig19.show()

conclusion:
1. Gender Distribution
Females represent the majority of customers at 70% compared to 30% for males, indicating that the mall primarily attracts a female customer base.
2. Age
The 56-65 age group represents the highest purchasing segment with 115 transactions, closely followed by the 18-25 group with 110, suggesting the mall attracts both senior and young customers. **The 26-35 age group shows the lowest purchasing activity**.
3. Payment Method
All three payment methods are relatively balanced, with females dominating across all options. Males strongly prefer Cash and avoid Credit Card, while Digital Wallet is the top choice for both genders.
4. Location
Egypt represents the largest customer base with 235 customers (44%), followed by UAE with 194, and USA with 100, indicating that the Egyptian market is the strongest.


5. High-Value Customers: Steven Knapp and Maria Weber lead in average spending per transaction (~$8,000+), making them ideal targets for **premium offers**.

 6. Most Loyal Customers: Brett Hodge tops total spending with ~$85,000 across 5 purchases, followed by Xavier Holloway, indicating strong loyalty and repeat buying behavior — **ideal candidates for loyalty programs.**

 7. 51.8% of customers are Regular buyers, with females dominating all segments. Converting the 27.7% one-time buyers — particularly males who show lower loyalty — into repeat customers remains a key growth opportunity.

**Recommendations:**

*Launch a loyalty program with points or rewards to
encourage one-time buyers to return.

*Target male customers with personalized offers and Cash-based promotions since they prefer Cash payments.

*Use email or digital wallet notifications to re-engage one-time buyers with exclusive deals.

**Product Analysis Insights:**
1. Sales by Category:

Electronics leads sales (136), while Clothing has the lowest demand (77).

2. Top Selling Products:

Rolex Sunglasses is the best-seller (32 transactions), followed by Nike Sneakers and Samsung Blender.

3. Top Brands by Sales Count:

Rolex and Samsung lead in number of transactions, followed by LG, indicating strong brand popularity across customers.

4. Brand Specialization:

Most brands are category-specific — Samsung dominates Home, Rolex leads Accessories, Nike focuses on Shoes, and H&M covers Clothing, reflecting clear brand-category alignmentز

**Promotion Analysis Insight:**

1. Promotions show limited effectiveness, with nearly equal split between promoted (49.5%) and non-promoted transactions. Both genders and all age groups purchase similarly with or without promotions, suggesting that customers' buying decisions are not significantly driven by promotions.

2. Promotions are heavily concentrated on Electronics and Home, despite Accessories being the best-selling category by volume. This suggests promotions are focused on high-revenue categories rather than high-demand ones.

3. Home receives the highest average discount (27%), yet it's not the top-selling category, while Clothing has the lowest discount (21%) and also the lowest sales — confirming that discount percentage alone does not drive purchasing decisions for most categories.

**Recommendation:**

*Consider redesigning promotion strategies by targeting underperforming age groups (26-35 and 36-45) with personalized offers to attract new buyers, while focusing promotions on low-performing categories like Clothing to drive incremental sales.

*Since Accessories already drives the highest sales without heavy promotion, consider redirecting promotional efforts to Clothing — the lowest performing category — to boost its demand and increase overall revenue diversification.
  
*Instead of increasing discounts, focus on product visibility and targeted marketing to boost underperforming categories like Clothing.

In [ ]:
df.head()

In [ ]:
df.promo_name.unique()

In [ ]:

df['amount_paid'] = df['total_amount'] * (1 - (df['discount_percentage'] / 100))


In [ ]:
df['total_cost'] = df['cost_price'] * df['quantity']
df['profit'] = df['amount_paid'] - df['total_cost']

In [ ]:
import plotly.express as px

# Aggregating data by branch
df_branch = df.groupby('branch_name')['profit'].sum().reset_index()

# Interactive Bar Chart
fig1 = px.bar(df_branch,
             x='branch_name',
             y='profit',
             title='Total Net Profit per Branch',
             labels={'profit': 'Net Profit ($)', 'branch_name': 'Branch Name'},
             color='profit',
             color_continuous_scale='RdYlGn') # Red to Green scale

fig1.show()

In [ ]:
import plotly.express as px

# 1. تجميع البيانات لحساب إجمالي المبيعات لكل فرع
# Aggregating total sales (total_amount) by branch_name
df_branch_sales = df.groupby('branch_name')['total_amount'].sum().reset_index()

# 2. ترتيب الفروع من الأكثر مبيعاً إلى الأقل (لجعل الرسم أكثر وضوحاً)
df_branch_sales = df_branch_sales.sort_values(by='total_amount', ascending=False)

# 3. إنشاء الرسم البياني التفاعلي
fig_total_sales = px.bar(df_branch_sales,
                         x='branch_name',
                         y='total_amount',
                         title='Total Sales Revenue by Branch', # عنوان إنجليزي واضح
                         labels={'total_amount': 'Total Revenue ($)', 'branch_name': 'Branch Name'},
                         color='total_amount', # تلوين الأعمدة بناءً على قيمة المبيعات
                         color_continuous_scale='GnBu', # تدرج لوني من الأخضر للأزرق
                         text='total_amount') # إظهار الرقم فوق كل عمود

# 4. تحسين تنسيق الأرقام والمظهر العام
fig_total_sales.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig_total_sales.update_layout(
    xaxis_title="Branch",
    yaxis_title="Total Amount ($)",
    plot_bgcolor='white', # خلفية بيضاء نظيفة
    showlegend=False
)

fig_total_sales.show()

In [ ]:
branch_fin = df.groupby('branch_name').agg({'total_cost':'sum', 'profit':'sum'}).reset_index()
fig = px.bar(branch_fin, x='branch_name', y=['total_cost', 'profit'],
             title='Total Cost vs Profit per Branch',
             barmode='group', color_discrete_sequence=['#EF553B', '#00CC96'])
fig.show()

In [ ]:
# توزيع البروموهات داخل كل فرع
fig_combined = px.bar(df,
                      x='branch_name',
                      color='promo_name',
                      title='توزيع العروض الترويجية لكل فرع',
                      barmode='group') # 'group' للمقارنة الجانبية أو 'stack' للتراكمي

fig_combined.show()

In [ ]:
import plotly.express as px

# 1. تجميع البيانات: إجمالي المباع مقابل إجمالي المخزون الحالي لكل فرع
inventory_analysis = df.groupby('branch_name').agg({
    'quantity': 'sum',          # إجمالي الكميات المباعة
    'stock_quantity': 'sum',    # إجمالي الكميات الموجودة في المخزن
    'profit': 'sum'             # الربح (لنستخدمه في حجم الفقاعة)
}).reset_index()

# 2. إنشاء الرسم البياني
fig_inventory = px.scatter(inventory_analysis,
                           x="quantity",
                           y="stock_quantity",
                           size="profit",
                           color="branch_name",
                           hover_name="branch_name",
                           title="<b>Inventory Efficiency: Sales Volume vs. Stock Level</b>",

                           text="branch_name", # إظهار اسم الفرع بجانب الفقاعة
                           size_max=60)

# 3. إضافة خط مرجعي (Trend Line) يدويًا أو تحسين الشكل
fig_inventory.update_layout(
    title_x=0.5,
    template="plotly_white",

)

fig_inventory.update_traces(textposition='top center')

fig_inventory.show()

In [ ]:
import plotly.express as px
import pandas as pd

# فلترة البيانات لتشمل فرع Mall_Branch_3 فقط
branch_df = df[df['branch_name'] == 'Mall_Branch_3']

In [ ]:
# تجميع البيانات حسب النوع
gender_counts = branch_df['gender'].value_counts().reset_index()
gender_counts.columns = ['gender', 'count']

# رسم بياني (Pie Chart)
fig_gender = px.pie(gender_counts, values='count', names='gender',
             title='Distribution of Customers by Gender in Mall_Branch_3',
             color_discrete_sequence=px.colors.qualitative.Pastel)
fig_gender.show()

In [ ]:
# تجميع البيانات حسب الفئة العمرية
age_group_counts = branch_df['age_group'].value_counts().reset_index()
age_group_counts.columns = ['age_group', 'count']

# رسم بياني (Bar Chart)
fig_age = px.bar(age_group_counts, x='age_group', y='count',
             title='Customer Age Groups in Mall_Branch_3',
             labels={'count': 'Number of Visits', 'age_group': 'Age Group'},
             color='count', color_continuous_scale='Viridis')
fig_age.show()

In [ ]:
# تجميع البيانات حسب الفئة وحساب إجمالي الكمية المباعة
category_analysis = branch_df.groupby('category')['quantity'].sum().reset_index().sort_values(by='quantity', ascending=False)

# رسم بياني (Horizontal Bar Chart)
fig_cat = px.bar(category_analysis, x='quantity', y='category',
             orientation='h',
             title='Best and Least Selling Categories in Mall_Branch_3',
             labels={'quantity': 'Total Quantity Sold', 'category': 'Product Category'},
             color='quantity', color_continuous_scale='Blues')

# تحديث الترتيب ليظهر الأكثر مبيعاً في الأعلى
fig_cat.update_layout(yaxis={'categoryorder':'total ascending'})
fig_cat.show()

In [ ]:
# تحويل عمود التاريخ إلى صيغة datetime إذا لم يكن كذلك
branch_df['sale_date'] = pd.to_datetime(branch_df['sale_date'])

# استخراج اسم يوم الأسبوع
branch_df['day_name'] = branch_df['sale_date'].dt.day_name()

# ترتيب الأيام بشكل منطقي (من السبت أو الاثنين حسب تفضيلك)
days_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

# تجميع المبيعات حسب اليوم
daily_sales = branch_df.groupby('day_name')['total_amount'].sum().reindex(days_order).reset_index()

# رسم بياني
fig_daily = px.line(daily_sales, x='day_name', y='total_amount',
              title='Sales Peak by Day of the Week (Mall_Branch_3)',
              markers=True, line_shape='spline',
              labels={'total_amount': 'Total Sales', 'day_name': 'Day of Week'})
fig_daily.show()

In [ ]:
# تجميع البيانات للأيام الثلاثة (الخميس، الجمعة، الأربعاء) للمقارنة
focus_days = ['Wednesday', 'Thursday', 'Friday']
comparison_df = branch_df[branch_df['day_name'].isin(focus_days)]

# حساب متوسط الخصم ومتوسط الربح لكل يوم
daily_analysis = comparison_df.groupby('day_name').agg({
    'profit': 'mean',
    'discount_percentage': 'mean',
    'total_amount': 'mean'
}).reindex(focus_days).reset_index()

# رسم بياني للمقارنة بين الربح ونسبة الخصم
import plotly.graph_objects as go

fig = go.Figure()
fig.add_trace(go.Bar(x=daily_analysis['day_name'], y=daily_analysis['profit'], name='Average Profit'))
fig.add_trace(go.Scatter(x=daily_analysis['day_name'], y=daily_analysis['discount_percentage'], name='Avg Discount %', yaxis='y2'))

fig.update_layout(
    title='Why is Thursday Higher? Profit vs Discount (Wed, Thu, Fri)',
    yaxis=dict(title='Average Profit'),
    yaxis2=dict(title='Discount %', overlaying='y', side='right')
)
fig.show()

In [ ]:
# حساب إجمالي الربح والمبيعات لكل نوع عرض
promo_analysis = branch_df.groupby('promo_name').agg({
    'total_amount': 'sum',
    'profit': 'sum',
    'quantity': 'sum'
}).reset_index().sort_values(by='profit', ascending=False)

# رسم بياني ثنائي المحاور (Bar للربح و Line للكمية)
import plotly.graph_objects as go

fig_promo = go.Figure()
fig_promo.add_trace(go.Bar(x=promo_analysis['promo_name'], y=promo_analysis['profit'], name='Total Profit'))
fig_promo.add_trace(go.Scatter(x=promo_analysis['promo_name'], y=promo_analysis['total_amount'], name='Total Sales', mode='lines+markers'))

fig_promo.update_layout(title='Promotion Performance: Profit vs Sales (Mall_Branch_3)')
fig_promo.show()

In [ ]:
# تصنيف العمليات: هل يوجد عرض أم لا؟
# نفترض أن العروض التي قيمتها 'None' أو 0 تعني بدون عرض
branch_df['is_promo'] = branch_df['promo_name'].apply(lambda x: 'No Promo' if x in ['None', 'No Promotion', 0] else 'With Promo')

# حساب إجمالي الأرباح لكل فئة
promo_vs_non = branch_df.groupby('is_promo')['profit'].sum().reset_index()

# رسم بياني للمقارنة
fig_compare = px.bar(promo_vs_non, x='is_promo', y='profit',
             title='Total Profit: Promotions vs. No Promotions',
             color='is_promo',
             color_discrete_map={'No Promo': '#2ecc71', 'With Promo': '#e74c3c'},
             text_auto=True)
fig_compare.show()

Conclusion & Strategic Insights
Mall Management System - Branch Analysis: Mall_Branch

Based on a comprehensive data analysis of Mall_Branch_3 and a comparative study of financial performance and customer behavior, the following insights were established:
1. The Promotion Paradox & Profitability Analysis

    Financial Discrepancy: Analysis reveals that Branch 3 is currently the lowest-performing branch in terms of sales and net profit, despite maintaining average operational costs (Total Cost).

    The Impact of Discounts: A critical discovery shows that non-promotional transactions (No Promo) yield significantly higher profits than discounted ones.

    Peak vs. Low Performance: Thursday was identified as the peak day for both sales and profit, characterized by a minimal reliance on promotions. Conversely, Friday and Wednesday recorded the lowest profit margins, even though Friday featured the highest frequency of promotional offers.

    Insight: Current promotional strategies are "cannibalizing" profit margins without driving a proportional increase in sales volume. Selling at Full Price in this branch has proven to be more financially viable.

2. Category & Product Performance

    Lead Category: The Electronics category is the top-performing segment, accounting for the highest volume of transactions and serving as the branch's primary revenue driver.

    Growth Opportunities: While Thursday drives high-value category sales, Friday tends to deplete inventory through deep discounts without generating a satisfactory Return on Investment (ROI).

3. Customer Demographic Profiling

    Gender Dominance: Data confirms that Females represent the primary purchasing power, showing higher transaction frequency and spending compared to males.

    The Golden Age Group: The branch's core customer base is concentrated within the 18–25 age bracket, clearly defining the branch as a youth-oriented shopping destination.

Strategic Recommendations

    Inventory Optimization: Increase the stock variety and floor space for Female-oriented products and Electronics, while reducing the inventory allocation for lower-performing male-specific items.

    The "Thursday" Strategy: Capitalize on Thursday's peak performance by launching new product arrivals or "Value-based" offers instead of "Discount-based" ones to maintain high margins.

    Friday Promotion Adjustment: We recommend halting high direct discounts on Friday for high-demand categories. Promotions should be restricted only to slow-moving inventory (High Stock Quantity) to clear warehouse space.

    Targeted Marketing: Direct social media and marketing campaigns toward the 18–25 age group to sustain the flow of the branch's most active customer segment.